tests/unit/s2n_server_alpn_extension_test.c

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_test.h"
+
+#include "tls/extensions/s2n_server_alpn.h"
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    const char *test_protocol_name = "chosen_protocol";
+    const uint8_t test_protocol_name_size = strlen(test_protocol_name);
+
+    /* Test should_send */
+    {
+        struct s2n_connection *conn;
+        EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
+
+        /* Should not send if protocol not set. Protocol not set by default. */
+        EXPECT_FALSE(s2n_server_alpn_extension.should_send(conn));
+
+        /* Should send if protocol set. */
+        EXPECT_MEMCPY_SUCCESS(conn->application_protocol, test_protocol_name, test_protocol_name_size);
+        EXPECT_TRUE(s2n_server_alpn_extension.should_send(conn));
+
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+    }
+
+    /* Test send */
+    {
+        struct s2n_connection *conn;
+        EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
+        EXPECT_MEMCPY_SUCCESS(conn->application_protocol, test_protocol_name, test_protocol_name_size);
+
+        struct s2n_stuffer stuffer;
+        EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&stuffer, 0));
+
+        EXPECT_SUCCESS(s2n_server_alpn_extension.send(conn, &stuffer));
+
+        /* Should have correct total size */
+        uint16_t total_size;
+        EXPECT_SUCCESS(s2n_stuffer_read_uint16(&stuffer, &total_size));
+        EXPECT_EQUAL(total_size, s2n_stuffer_data_available(&stuffer));
```

Comment:
What is `total_size` here referring to?

Reply:
Total size of the protocol name list. I can make that clearer.